packages/core/core/src/loadParcelPlugin.js

```diff
@@ -10,35 +9,100 @@ import nullthrows from 'nullthrows';
 
+const NODE_MODULES = `${path.sep}node_modules${path.sep}`;
+
 export default async function loadPlugin<T>(
-  fs: FileSystem,
-  packageManager: PackageManager,
   pluginName: PackageName,
-  resolveFrom: FilePath,
+  configPath: FilePath,
   keyPath: string,
-  shouldAutoInstall: boolean,
+  options: ParcelOptions,
 ): Promise<{|plugin: T, version: Semver|}> {
+  let resolveFrom = configPath;
+  let range;
+  if (resolveFrom.includes(NODE_MODULES)) {
+    let configPkg = await loadConfig(options.inputFS, resolveFrom, [
+      'package.json',
+    ]);
+    if (
+      configPkg != null &&
+      configPkg.config.dependencies?.[pluginName] == null
+    ) {
+      // If not in the config's dependencies, the plugin will be auto installed with
+      // the version declared in "parcelDependencies".
+      range = configPkg.config.parcelDependencies?.[pluginName];
+
+      if (range == null) {
+        let contents = await options.inputFS.readFile(
+          configPkg.files[0].filePath,
+          'utf8',
+        );
+        throw new ThrowableDiagnostic({
+          diagnostic: {
+            message: `Could not determine version of ${pluginName} in ${path.relative(
+              process.cwd(),
+              resolveFrom,
+            )}. Either include it in "dependencies" or "parcelDependencies".`,
+            origin: '@parcel/core',
+            filePath: configPkg.files[0].filePath,
+            language: 'json5',
+            codeFrame:
+              configPkg.config.dependencies ||
+              configPkg.config.parcelDependencies
+                ? {
+                    code: contents,
+                    codeHighlights: generateJSONCodeHighlights(contents, [
+                      {
+                        key: configPkg.config.parcelDependencies
+                          ? '/parcelDependencies'
+                          : '/dependencies',
+                        type: 'key',
+                      },
+                    ]),
+                  }
+                : undefined,
+          },
+        });
+      }
+
+      // Resolve from project root if not in the config's dependencies.
+      resolveFrom = path.join(options.projectRoot, 'index');
+    }
+  }
+
   let resolved, pkg;
   try {
-    ({resolved, pkg} = await packageManager.resolve(pluginName, resolveFrom, {
-      shouldAutoInstall,
-    }));
+    ({resolved, pkg} = await options.packageManager.resolve(
+      pluginName,
+      resolveFrom,
+      {
+        shouldAutoInstall: options.shouldAutoInstall,
+        range,
+      },
+    ));
   } catch (err) {
-    let configContents = await fs.readFile(resolveFrom, 'utf8');
+    if (err.code !== 'MODULE_NOT_FOUND') {
+      throw err;
+    }
+
+    let configContents = await options.inputFS.readFile(configPath, 'utf8');
     let alternatives = await findAlternativeNodeModules(
-      fs,
+      options.inputFS,
       pluginName,
       path.dirname(resolveFrom),
     );
     throw new ThrowableDiagnostic({
```

Comment:
(Not really specific to the changes, but this reminds me of how `Cannot find Parcel plugin` would be printed even when Yarn failed, e.g. because there is no such package: https://github.com/parcel-bundler/parcel/issues/5750)